In [9]:
import os
import pandas as pd
import numpy as np
pd.set_option("display.max_colwidth", 2)

os.chdir("/home/coinse/Workspace/SBFL")

# Load bug info

In [10]:
from scripts.util import load_bug_data, find_covmat_func, find_covmat_idx
from functools import reduce
from typing import Set

pid, vid = "wireshark", 4

bug = load_bug_data(pid, vid)

print(f"{bug.buggy_files=}")
print(f"{bug.buggy_lines=}")
print("bug.coverage_df=")
display(bug.coverage_df)
print(f"{bug.failing_tests=}")
print(f"{bug.failing_tests_GT=}")
print("bug.functions={", end="")
for func in list(bug.functions)[:5]:
    print(f"{func},", end=" ")
print(", ...}")
print(f"{bug.id=}")
print(f"{bug.num_test_cases=}")
print(f"{bug.passing_tests=}")
print(f"{bug.path_to_cov_df=}")
print(f"{bug.path_to_patch=}")
print(f"{bug.path_to_test_results=}")
print("bug.source_files=", end="")
for src in list(bug.source_files)[:5]:
    print(f"{src},", end=" ")
print(", ...}")

fault_idxs = sorted(bug.buggy_lines)
# fault_idx2meth = {}
fault_meth2idx = {}
err_msg = ""
for fault_idx in fault_idxs:
    succ_func, ret_func = find_covmat_func(
        pid, bug.coverage_df, fault_idx, debug=False
    )
    if succ_func:
        succ_idx, ret_idx = find_covmat_idx(
            pid, bug.coverage_df, fault_idx, debug=False
        )
        if succ_idx:
            if ret_func in fault_meth2idx:
                fault_meth2idx[ret_func].add(ret_idx)
            else:
                fault_meth2idx[ret_func] = {ret_idx}
        else:
            print(f"{ret_idx}")
            if ret_func not in fault_meth2idx:
                fault_meth2idx[ret_func] = set()
    else:
        print(f"{ret_func}")
# fault_meths: Set = reduce(
#     lambda x, y: x | set([y]), fault_idx2meth.values(), set()
# )
# fault_meth2idx = {
#     v: {idx for idx in fault_idx2meth if fault_idx2meth[idx] == v}
#     for v in fault_meths
# }
fault_meths = set(fault_meth2idx.keys())
for meth in fault_meth2idx:
    print(f"{meth=}: {fault_meth2idx[meth]}")
# print(f"{fault_meths=}")


bug.buggy_files={'epan/crypt/dot11decrypt.c'}
bug.buggy_lines={('epan/crypt/dot11decrypt.c', 330), ('epan/crypt/dot11decrypt.c', 1603), ('epan/crypt/dot11decrypt.c', 1590), ('epan/crypt/dot11decrypt.c', 1602), ('epan/crypt/dot11decrypt.c', 1599), ('epan/crypt/dot11decrypt.c', 1589), ('epan/crypt/dot11decrypt.c', 332), ('epan/crypt/dot11decrypt.c', 1601), ('epan/crypt/dot11decrypt.c', 331), ('epan/crypt/dot11decrypt.c', 1591), ('epan/crypt/dot11decrypt.c', 1604), ('epan/crypt/dot11decrypt.c', 1600)}
bug.coverage_df=


28  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   461   
                                                                                  177   461   
                                                                                  199   461   
                                                                                  203   461   
                                                                                  205   461   
...                                                                                     ...   
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1720  461   
                                                                                  1721  461   
                                                                                  1723  461   
                                                                                  1724  461   
                                                                                  1725  461   

                                                                                          6  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   216   
                                                                                  177   216   
                                                                                  199   216   
                                                                                  203   216   
                                                                                  205   216   
...                                                                                     ...   
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1720  216   
                                                                                  1721  216   
                                                                                  1723  216   
                                                                                  1724  216   
                                                                                  1725  216   

                                                                                          5  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   214   
                                                                                  177   214   
                                                                                  199   214   
                                                                                  203   214   
                                                                                  205   214   
...                                                                                     ...   
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1720  214   
                                                                                  1721  214   
                                                                                  1723  214   
                                                                                  1724  214   
                                                                                  1725  214   

                                                                                         15  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   368   
                                                                                  177   368   
 

bug.failing_tests={'3'}
bug.failing_tests_GT={'3'}
bug.functions={('/home/workspace/epan/packet.c', 'find_dissector_table'), ('/home/workspace/capinfos.c', 'process_cap_file'), ('/home/workspace/epan/ftypes/ftype-integer.c', 'set_uinteger64'), ('/home/workspace/epan/dissectors/packet-starteam.c', 'dissect_starteam_heur'), ('/home/workspace/epan/dissectors/packet-eap.c', 'dissect_eap_identity_wlan'), , ...}
bug.id='wireshark-4'
bug.num_test_cases=34
bug.passing_tests={'6', '21', '16', '25', '2', '23', '5', '24', '8', '34', '33', '29', '15', '31', '17', '28', '32', '7', '11', '12', '1', '30', '13', '4', '9', '26', '20', '14', '19', '10', '27', '22', '18'}
bug.path_to_cov_df='/home/coinse/Workspace/defects4cpp/coverage/wireshark_4_buggy/coverage_df.pkl'
bug.path_to_patch='/home/coinse/Workspace/defects4cpp/defects4cpp/taxonomy/wireshark/patch/0004-buggy.patch'
bug.path_to_test_results='/home/coinse/Workspace/defects4cpp/coverage/wireshark_4_buggy/test_results.json'
bug.source_files=/home/

# Get spectra

In [11]:
covmat = bug.coverage_df
pass_tcs = bug.passing_tests
fail_tcs = bug.failing_tests

covmat_bool = covmat.astype(bool)
P = len(pass_tcs)
F = len(fail_tcs)
e_p = covmat_bool[pass_tcs].sum(axis=1)
e_f = covmat_bool[fail_tcs].sum(axis=1)
n_p = P - e_p
n_f = F - e_f
spectra = pd.DataFrame(
    {
        "line": covmat.index.to_list(),
        "e_p": e_p,
        "e_f": e_f,
        "n_p": n_p,
        "n_f": n_f,
    }
)

display(spectra)


line  \
file                                                  function                    line                                                                                              
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 174)   
                                                                                  177   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 177)   
                                                                                  199   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 199)   
                                                                                  203   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 203)   
                                                                                  205   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 205)   
...                                                                                                                                                                           ...   
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1720  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1720)               
                                                                                  1721  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1721)               
                                                                                  1723  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1723)               
                                                                                  1724  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1724)               
                                                                                  1725  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1725)               

                                                                                        e_p  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   33    
                                                                                  177   33    
                                                                                  199   33    
                                                                                  203   33    
                                                                                  205   33    
...                                                                                     ..    
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1720  33    
                                                                                  1721  33    
                                                                                  1723  33    
                                                                                  1724  33    
                                                                                  1725  33    

                                                                                        e_f  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   1     
                                                                                  177   1     
                                                                                  199   1     
                                                                                  203   1     
             

# Compute suspiciousness scores

In [12]:
def ochiai(e_f, e_p, n_f, n_p):
    return e_f / np.sqrt(e_p + e_f) / np.sqrt(e_f + n_f) if e_f else 0.0


def op2(e_f, e_p, n_f, n_p):
    return e_f - e_p / (e_p + n_p + 1)

spectra["Sop2"] = spectra.apply(
    lambda row: op2(row.e_f, row.e_p, row.n_f, row.n_p), axis=1
)
spectra["Sochiai"] = spectra.apply(
    lambda row: ochiai(row.e_f, row.e_p, row.n_f, row.n_p), axis=1
)
spectra["Rop2"] = spectra.Sop2.rank(ascending=False, method="max")
spectra["Rochiai"] = spectra.Sochiai.rank(ascending=False, method="max")


# Line level SBFL

In [13]:
for fault_meth in fault_meth2idx:
    fault_idxs_of_meth = fault_meth2idx[fault_meth]
    print("fault meth:", fault_meth)
    for fault_idx in fault_idxs_of_meth:
        print("fault idx:", tuple(fault_idx))
    print("SBFL score of faulty method:")
    display(spectra.loc[fault_idxs_of_meth].reset_index(drop=True))

print("Top 5 SBFL scores:")
print("Op2:")
display(
    spectra.sort_values("Sop2", ascending=False).head(5).reset_index(drop=True)
)
print("Ochiai:")
display(
    spectra.sort_values("Sochiai", ascending=False)
    .head(5)
    .reset_index(drop=True)
)


fault meth: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptCopyKey')
fault idx: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptCopyKey', 330)
fault idx: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptCopyKey', 331)
SBFL score of faulty method:


line  e_p  \
0  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey, 330)  31    
1  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey, 331)  31    

   e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  
0  1    2    0    0.088235  0.176777  16902.0  16902.0  
1  1    2    0    0.088235  0.176777  16902.0  16902.0

fault meth: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptRsna4WHandshake')
fault idx: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptRsna4WHandshake', 1604)
fault idx: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptRsna4WHandshake', 1600)
fault idx: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptRsna4WHandshake', 1590)
fault idx: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptRsna4WHandshake', 1603)
fault idx: ('/home/workspace/epan/crypt/dot11decrypt.c', 'Dot11DecryptRsna4WHandshake', 1602)
SBFL score of faulty method:


line  \
0  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake, 1604)   
1  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake, 1600)   
2  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake, 1590)   
3  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake, 1603)   
4  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake, 1602)   

   e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  
0  31   1    2    0    0.088235  0.176777  16902.0  16902.0  
1  31   1    2    0    0.088235  0.176777  16902.0  16902.0  
2  31   1    2    0    0.088235  0.176777  16902.0  16902.0  
3  31   1    2    0    0.088235  0.176777  16902.0  16902.0  
4  31   1    2    0    0.088235  0.176777  16902.0  16902.0

Top 5 SBFL scores:
Op2:


line  e_p  \
0  (/home/workspace/epan/prefs.c, register_preference, 1039)             2     
1  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name, 7392)  7     
2  (/home/workspace/epan/proto.c, find_protocol_by_id, 7298)             7     
3  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name, 7385)  7     
4  (/home/workspace/epan/proto.c, proto_get_id, 7315)                    7     

   e_f  n_p  n_f      Sop2   Sochiai  Rop2  Rochiai  
0  1    31   0    0.941176  0.577350  1.0   1.0      
1  1    26   0    0.794118  0.353553  13.0  13.0     
2  1    26   0    0.794118  0.353553  13.0  13.0     
3  1    26   0    0.794118  0.353553  13.0  13.0     
4  1    26   0    0.794118  0.353553  13.0  13.0

Ochiai:


line  e_p  \
0  (/home/workspace/epan/prefs.c, register_preference, 1039)             2     
1  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name, 7392)  7     
2  (/home/workspace/epan/proto.c, find_protocol_by_id, 7298)             7     
3  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name, 7385)  7     
4  (/home/workspace/epan/proto.c, proto_get_id, 7315)                    7     

   e_f  n_p  n_f      Sop2   Sochiai  Rop2  Rochiai  
0  1    31   0    0.941176  0.577350  1.0   1.0      
1  1    26   0    0.794118  0.353553  13.0  13.0     
2  1    26   0    0.794118  0.353553  13.0  13.0     
3  1    26   0    0.794118  0.353553  13.0  13.0     
4  1    26   0    0.794118  0.353553  13.0  13.0

# Method level SBFL

In [14]:
spectra["file"] = spectra.index.map(lambda x: x[0])
spectra["func"] = spectra.index.map(lambda x: x[1])
spectra["line"] = spectra.index.map(lambda x: x[2])
# (file, func) is meth
spectra["meth"] = spectra.apply(lambda row: (row.file, row.func), axis=1)
spectra.reset_index(drop=True, inplace=True)
display(spectra)

line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0      174   33   1    0    0    0.029412  0.171499  79913.0  79913.0   
1      177   33   1    0    0    0.029412  0.171499  79913.0  79913.0   
2      199   33   1    0    0    0.029412  0.171499  79913.0  79913.0   
3      203   33   1    0    0    0.029412  0.171499  79913.0  79913.0   
4      205   33   1    0    0    0.029412  0.171499  79913.0  79913.0   
...    ...   ..  ..   ..   ..         ...       ...      ...      ...   
96795  1720  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
96796  1721  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
96797  1723  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
96798  1724  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
96799  1725  33   1    0    0    0.029412  0.171499  79913.0  79913.0   

                                                        file  \
0      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
1      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
2      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
3      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
4      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
...                                                      ...   
96795  /home/workspace/epan/dissectors/packet-isns.c           
96796  /home/workspace/epan/dissectors/packet-isns.c           
96797  /home/workspace/epan/dissectors/packet-isns.c           
96798  /home/workspace/epan/dissectors/packet-isns.c           
96799  /home/workspace/epan/dissectors/packet-isns.c           

                              func  \
0      proto_register_gluster_pmap   
1      proto_register_gluster_pmap   
2      proto_register_gluster_pmap   
3      proto_register_gluster_pmap   
4      proto_register_gluster_pmap   
...                            ...   
96795  proto_reg_handoff_isns        
96796  proto_reg_handoff_isns        
96797  proto_reg_handoff_isns        
96798  proto_reg_handoff_isns        
96799  proto_reg_handoff_isns        

                                                                                       meth  
0      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
1      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
2      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
3      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
4      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
...                                                                                     ...  
96795  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               
96796  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               
96797  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               
96798  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               
96799  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               

[96800 rows x 12 columns]

## Max aggregation

In [15]:
print("[Op2]")
spectra_max_op2 = spectra.copy().sort_values(
    ["Sop2", "meth", "line"], ascending=False
)
spectra_max_op2.drop_duplicates(subset="meth", keep="first", inplace=True)
spectra_max_op2 = spectra_max_op2[
    ["meth", "line", "e_p", "e_f", "n_p", "n_f", "Sop2"]
]
spectra_max_op2["Rop2"] = spectra_max_op2.Sop2.rank(
    ascending=False, method="max"
)
print("Faulty method:")
display(
    spectra_max_op2.loc[spectra_max_op2.meth.isin(fault_meths)].reset_index(
        drop=True
    )
)
print("Top 20 methods:")
display(spectra_max_op2.head(20).reset_index(drop=True))
print()
print("[Ochiai]")
spectra_max_ochiai = spectra.copy().sort_values(
    ["Sochiai", "meth", "line"], ascending=False
)
spectra_max_ochiai.drop_duplicates(subset="meth", keep="first", inplace=True)
spectra_max_ochiai = spectra_max_ochiai[
    ["meth", "line", "e_p", "e_f", "n_p", "n_f", "Sochiai"]
]
spectra_max_ochiai["Rochiai"] = spectra_max_ochiai.Sochiai.rank(
    ascending=False, method="max"
)
print("Faulty method:")
display(
    spectra_max_ochiai.loc[
        spectra_max_ochiai.meth.isin(fault_meths)
    ].reset_index(drop=True)
)
print("Top 20 methods:")
display(spectra_max_ochiai.head(20).reset_index(drop=True))


[Op2]
Faulty method:


meth  \
0  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
1  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)           

   line  e_p  e_f  n_p  n_f      Sop2    Rop2  
0  1759  31   1    2    0    0.088235  1068.0  
1  369   31   1    2    0    0.088235  1068.0

Top 20 methods:


meth  \
0   (/home/workspace/epan/prefs.c, register_preference)                        
1   (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)             
2   (/home/workspace/epan/proto.c, proto_get_id)                               
3   (/home/workspace/epan/proto.c, find_protocol_by_id)                        
4   (/home/workspace/epan/addr_resolv.c, parse_ether_address)                  
5   (/home/workspace/epan/oids.c, check_num_oid)                               
6   (/home/workspace/epan/packet.c, find_matching_proto_name)                  
7   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsnaPwd2PskStep)   
8   (/home/workspace/epan/wmem/wmem_map.c, wmem_strong_hash)                   
9   (/home/workspace/epan/proto.c, tmp_fld_check_assert)                       
10  (/home/workspace/epan/proto.c, proto_register_field_init)                  
11  (/home/workspace/epan/proto.c, proto_register_field_common)                
12  (/home/workspace/epan/proto.c, proto_register_field_array)                 
13  (/home/workspace/epan/addr_resolv.c, fgetline)                             
14  (/home/workspace/epan/wmem/wmem_core.c, wmem_alloc)                        
15  (/home/workspace/epan/packet.c, dissector_compare_filter_name)             
16  (/home/workspace/epan/packet.c, dissector_add_for_decode_as)               
17  (/usr/include/x86_64-linux-gnu/sys/stat.h, nan)                            
18  (/usr/include/x86_64-linux-gnu/gpg-error.h, nan)                           
19  (/usr/include/gcrypt.h, nan)                                               

    line  e_p  e_f  n_p  n_f      Sop2    Rop2  
0   1039  2    1    31   0    0.941176  1.0     
1   7392  7    1    26   0    0.794118  4.0     
2   7315  7    1    26   0    0.794118  4.0     
3   7309  7    1    26   0    0.794118  4.0     
4   1388  10   1    23   0    0.705882  5.0     
5   893   14   1    19   0    0.588235  6.0     
6   3323  22   1    11   0    0.352941  8.0     
7   2427  22   1    11   0    0.352941  8.0     
8   428   24   1    9    0    0.294118  9.0     
9   8347  29   1    4    0    0.147059  14.0    
10  8560  29   1    4    0    0.147059  14.0    
11  7581  29   1    4    0    0.147059  14.0    
12  7615  29   1    4    0    0.147059  14.0    
13  607   29   1    4    0    0.147059  14.0    
14  34    30   1    3    0    0.117647  17.0    
15  2002  30   1    3    0    0.117647  17.0    
16  2070  30   1    3    0    0.117647  17.0    
17  469   31   1    2    0    0.088235  1068.0  
18  969   31   1    2    0    0.088235  1068.0  
19  160   31   1    2    0    0.088235  1068.0


[Ochiai]
Faulty method:


meth  \
0  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
1  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)           

   line  e_p  e_f  n_p  n_f   Sochiai  Rochiai  
0  1759  31   1    2    0    0.176777  1068.0   
1  369   31   1    2    0    0.176777  1068.0

Top 20 methods:


meth  \
0   (/home/workspace/epan/prefs.c, register_preference)                        
1   (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)             
2   (/home/workspace/epan/proto.c, proto_get_id)                               
3   (/home/workspace/epan/proto.c, find_protocol_by_id)                        
4   (/home/workspace/epan/addr_resolv.c, parse_ether_address)                  
5   (/home/workspace/epan/oids.c, check_num_oid)                               
6   (/home/workspace/epan/packet.c, find_matching_proto_name)                  
7   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsnaPwd2PskStep)   
8   (/home/workspace/epan/wmem/wmem_map.c, wmem_strong_hash)                   
9   (/home/workspace/epan/proto.c, tmp_fld_check_assert)                       
10  (/home/workspace/epan/proto.c, proto_register_field_init)                  
11  (/home/workspace/epan/proto.c, proto_register_field_common)                
12  (/home/workspace/epan/proto.c, proto_register_field_array)                 
13  (/home/workspace/epan/addr_resolv.c, fgetline)                             
14  (/home/workspace/epan/wmem/wmem_core.c, wmem_alloc)                        
15  (/home/workspace/epan/packet.c, dissector_compare_filter_name)             
16  (/home/workspace/epan/packet.c, dissector_add_for_decode_as)               
17  (/usr/include/x86_64-linux-gnu/sys/stat.h, nan)                            
18  (/usr/include/x86_64-linux-gnu/gpg-error.h, nan)                           
19  (/usr/include/gcrypt.h, nan)                                               

    line  e_p  e_f  n_p  n_f   Sochiai  Rochiai  
0   1039  2    1    31   0    0.577350  1.0      
1   7392  7    1    26   0    0.353553  4.0      
2   7315  7    1    26   0    0.353553  4.0      
3   7309  7    1    26   0    0.353553  4.0      
4   1388  10   1    23   0    0.301511  5.0      
5   893   14   1    19   0    0.258199  6.0      
6   3323  22   1    11   0    0.208514  8.0      
7   2427  22   1    11   0    0.208514  8.0      
8   428   24   1    9    0    0.200000  9.0      
9   8347  29   1    4    0    0.182574  14.0     
10  8560  29   1    4    0    0.182574  14.0     
11  7581  29   1    4    0    0.182574  14.0     
12  7615  29   1    4    0    0.182574  14.0     
13  607   29   1    4    0    0.182574  14.0     
14  34    30   1    3    0    0.179605  17.0     
15  2002  30   1    3    0    0.179605  17.0     
16  2070  30   1    3    0    0.179605  17.0     
17  469   31   1    2    0    0.176777  1068.0   
18  969   31   1    2    0    0.176777  1068.0   
19  160   31   1    2    0    0.176777  1068.0

## Line2Method voting (weight: sbfl score / dense rank)

In [16]:
print("[Op2]")
ballot_op2 = {}
spectra_max_op2 = spectra.copy().sort_values(
    ["Sop2", "meth", "line"], ascending=False
)
spectra_max_op2["Rop2D"] = spectra_max_op2.Sop2.rank(
    ascending=False, method="dense"
)
spectra_max_op2["vote"] = spectra_max_op2.Sop2 / spectra_max_op2.Rop2D
for row in spectra_max_op2.itertuples():
    if row.meth not in ballot_op2:
        ballot_op2[row.meth] = row.vote
    else:
        ballot_op2[row.meth] += row.vote
L2M_op2 = pd.DataFrame(
    {"meth": list(ballot_op2.keys()), "vote": list(ballot_op2.values())}
)
L2M_op2["rank"] = L2M_op2.vote.rank(ascending=False, method="max")
L2M_op2.sort_values("rank", inplace=True)
L2M_op2.reset_index(drop=True, inplace=True)
for fault_meth in fault_meths:
    print("Faulty method:")
    display(L2M_op2.loc[L2M_op2.meth == fault_meth].reset_index(drop=True))
    print("Voters:")
    display(
        spectra_max_op2.loc[spectra_max_op2.meth == fault_meth].reset_index(
            drop=True
        )
    )
print("Top 5 methods:")
display(L2M_op2.head(5).reset_index(drop=True))
for idx, top_meth in enumerate(L2M_op2.head(5).meth.tolist(), 1):
    print(f"Top {idx}:")
    display(L2M_op2.loc[L2M_op2.meth == top_meth].reset_index(drop=True))
    print("Voters:")
    display(
        spectra_max_op2.loc[spectra_max_op2.meth == top_meth].reset_index(
            drop=True
        )
    )
print()
print("[Ochiai]")
ballot_ochiai = {}
spectra_max_ochiai = spectra.copy().sort_values(
    ["Sochiai", "meth", "line"], ascending=False
)
spectra_max_ochiai["RochiaiD"] = spectra_max_ochiai.Sochiai.rank(
    ascending=False, method="dense"
)
spectra_max_ochiai["vote"] = (
    spectra_max_ochiai.Sochiai / spectra_max_ochiai.RochiaiD
)
for row in spectra_max_ochiai.itertuples():
    if row.meth not in ballot_ochiai:
        ballot_ochiai[row.meth] = row.vote
    else:
        ballot_ochiai[row.meth] += row.vote
L2M_ochiai = pd.DataFrame(
    {"meth": list(ballot_ochiai.keys()), "vote": list(ballot_ochiai.values())}
)
L2M_ochiai["rank"] = L2M_ochiai.vote.rank(ascending=False, method="max")
L2M_ochiai.sort_values("rank", inplace=True)
L2M_ochiai.reset_index(drop=True, inplace=True)
for fault_meth in fault_meths:
    print("Faulty method:")
    display(
        L2M_ochiai.loc[L2M_ochiai.meth == fault_meth].reset_index(drop=True)
    )
    print("Voters:")
    display(
        spectra_max_ochiai.loc[
            spectra_max_ochiai.meth == fault_meth
        ].reset_index(drop=True)
    )
print("Top 5 methods:")
display(L2M_ochiai.head(5).reset_index(drop=True))
for idx, top_meth in enumerate(L2M_ochiai.head(5).meth.tolist(), 1):
    print(f"Top {idx}:")
    display(L2M_ochiai.loc[L2M_ochiai.meth == top_meth].reset_index(drop=True))
    print("Voters:")
    display(
        spectra_max_ochiai.loc[spectra_max_ochiai.meth == top_meth].reset_index(
            drop=True
        )
    )


[Op2]
Faulty method:


meth      vote  \
0  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)  0.220588   

    rank  
0  260.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0   369   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1   355   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2   354   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3   353   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4   352   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
5   351   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
6   350   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
7   349   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
8   348   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
9   347   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
10  343   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
11  342   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
12  341   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
13  334   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
14  331   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
15  330   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
16  329   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
17  328   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
18  327   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
19  326   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
20  325   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
21  323   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
22  322   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
23  321   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
24  319   31   1    2    0    0.088235  0.176777  16902.0  16902.0   

                                         file                 func  \
0   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
1   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
2   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
3   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
4   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
5   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
6   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
7   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
8   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
9   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
10  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
11  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
12  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
13  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
14  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
15  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
16  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
17  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
18  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
19  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
20  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
21  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
22  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
23  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
24  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   

                                                                meth  Rop2D  \
0   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)  10.0    
1   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)  10.0    
2   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)  10.0    
3   (/home/workspace/epan/crypt/dot11decrypt.c

Faulty method:


meth  \
0  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   

   vote  rank  
0  0.9   26.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    1759  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    1725  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2    1724  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3    1722  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4    1715  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
97   1477  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
98   1476  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
99   1475  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
100  1474  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
101  1467  31   1    2    0    0.088235  0.176777  16902.0  16902.0   

                                          file                         func  \
0    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
1    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
2    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
3    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
4    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
..                                         ...                          ...   
97   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
98   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
99   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
100  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
101  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   

                                                                         meth  \
0    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
1    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
2    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
3    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
4    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
..                                                                        ...   
97   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
98   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
99   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
100  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
101  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   

     Rop2D      vote  
0    10.0   0.008824  
1    10.0   0.008824  
2    10.0   0.008824  
3    10.0   0.008824  
4    10.0   0.008824  
..    ...        ...  
97   10.0   0.008824  
98   10.0   0.008824  
99   10.0   0.008824  
100  10.0   0.008824  
101  10.0   0.008824  

[102 rows x 14 columns]

Top 5 methods:


meth  \
0  (/home/workspace/epan/proto.c, find_protocol_by_id)                                   
1  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)                        
2  (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)       
3  (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
4  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)                    

       vote  rank  
0  1.985483  1.0   
1  1.985294  2.0   
2  1.863235  3.0   
3  1.788914  4.0   
4  1.487485  5.0

Top 1:


meth      vote  rank
0  (/home/workspace/epan/proto.c, find_protocol_by_id)  1.985483  1.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0  7309  7    1    26   0    0.794118  0.353553  13.0     13.0      
1  7306  7    1    26   0    0.794118  0.353553  13.0     13.0      
2  7302  7    1    26   0    0.794118  0.353553  13.0     13.0      
3  7300  7    1    26   0    0.794118  0.353553  13.0     13.0      
4  7298  7    1    26   0    0.794118  0.353553  13.0     13.0      
5  7307  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
6  7305  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   

                           file                 func  \
0  /home/workspace/epan/proto.c  find_protocol_by_id   
1  /home/workspace/epan/proto.c  find_protocol_by_id   
2  /home/workspace/epan/proto.c  find_protocol_by_id   
3  /home/workspace/epan/proto.c  find_protocol_by_id   
4  /home/workspace/epan/proto.c  find_protocol_by_id   
5  /home/workspace/epan/proto.c  find_protocol_by_id   
6  /home/workspace/epan/proto.c  find_protocol_by_id   

                                                  meth  Rop2D      vote  
0  (/home/workspace/epan/proto.c, find_protocol_by_id)  2.0    0.397059  
1  (/home/workspace/epan/proto.c, find_protocol_by_id)  2.0    0.397059  
2  (/home/workspace/epan/proto.c, find_protocol_by_id)  2.0    0.397059  
3  (/home/workspace/epan/proto.c, find_protocol_by_id)  2.0    0.397059  
4  (/home/workspace/epan/proto.c, find_protocol_by_id)  2.0    0.397059  
5  (/home/workspace/epan/proto.c, find_protocol_by_id)  12.0   0.002451  
6  (/home/workspace/epan/proto.c, find_protocol_by_id)  13.0  -0.002262

Top 2:


meth      vote  \
0  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)  1.985294   

   rank  
0  2.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai  Rop2  Rochiai  \
0  7392  7    1    26   0    0.794118  0.353553  13.0  13.0      
1  7390  7    1    26   0    0.794118  0.353553  13.0  13.0      
2  7389  7    1    26   0    0.794118  0.353553  13.0  13.0      
3  7387  7    1    26   0    0.794118  0.353553  13.0  13.0      
4  7385  7    1    26   0    0.794118  0.353553  13.0  13.0      

                           file                            func  \
0  /home/workspace/epan/proto.c  proto_get_protocol_filter_name   
1  /home/workspace/epan/proto.c  proto_get_protocol_filter_name   
2  /home/workspace/epan/proto.c  proto_get_protocol_filter_name   
3  /home/workspace/epan/proto.c  proto_get_protocol_filter_name   
4  /home/workspace/epan/proto.c  proto_get_protocol_filter_name   

                                                             meth  Rop2D  \
0  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)  2.0     
1  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)  2.0     
2  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)  2.0     
3  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)  2.0     
4  (/home/workspace/epan/proto.c, proto_get_protocol_filter_name)  2.0     

       vote  
0  0.397059  
1  0.397059  
2  0.397059  
3  0.397059  
4  0.397059

Top 3:


meth  \
0  (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   

       vote  rank  
0  1.863235  3.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    3026  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    3023  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2    2985  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3    2980  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4    2978  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
208  2100  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
209  2099  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
210  2098  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
211  2096  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
212  2338  4    0    29   1   -0.117647  0.000000  91069.0  96800.0   

                                                            file  \
0    /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
1    /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
2    /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
3    /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
4    /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
..                                                           ...   
208  /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
209  /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
210  /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
211  /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   
212  /home/workspace/epan/dissectors/packet-ieee80211-radiotap.c   

                 func  \
0    dissect_radiotap   
1    dissect_radiotap   
2    dissect_radiotap   
3    dissect_radiotap   
4    dissect_radiotap   
..                ...   
208  dissect_radiotap   
209  dissect_radiotap   
210  dissect_radiotap   
211  dissect_radiotap   
212  dissect_radiotap   

                                                                                meth  \
0    (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
1    (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
2    (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
3    (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
4    (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
..                                                                               ...   
208  (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
209  (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
210  (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
211  (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   
212  (/home/workspace/epan/dissectors/packet-ieee80211-radiotap.c, dissect_radiotap)   

     Rop2D      vote  
0    10.0   0.008824  
1    10.0   0.008824  
2    10.0   0.008824  
3    10.0   0.008824  
4    10.0   0.008824  
..    ...        ...  
208  10.0   0.008824  
209  10.0   0.008824  
210  10.0   0.008824  
211  10.0   0.008824  
212  16.0  -0.007353  

[213 rows x 14 columns]

Top 4:


meth  \
0  (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   

       vote  rank  
0  1.788914  4.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    1224  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    1222  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2    1221  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3    1217  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4    1216  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
199  459   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
200  458   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
201  457   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
202  455   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
203  901   1    0    32   1   -0.029412  0.000000  85030.0  96800.0   

                                                         file  \
0    /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
1    /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
2    /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
3    /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
4    /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
..                                                        ...   
199  /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
200  /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
201  /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
202  /home/workspace/epan/dissectors/packet-ieee80211-radio.c   
203  /home/workspace/epan/dissectors/packet-ieee80211-radio.c   

                        func  \
0    dissect_wlan_radio_phdr   
1    dissect_wlan_radio_phdr   
2    dissect_wlan_radio_phdr   
3    dissect_wlan_radio_phdr   
4    dissect_wlan_radio_phdr   
..                       ...   
199  dissect_wlan_radio_phdr   
200  dissect_wlan_radio_phdr   
201  dissect_wlan_radio_phdr   
202  dissect_wlan_radio_phdr   
203  dissect_wlan_radio_phdr   

                                                                                    meth  \
0    (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
1    (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
2    (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
3    (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
4    (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
..                                                                                   ...   
199  (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
200  (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
201  (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
202  (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   
203  (/home/workspace/epan/dissectors/packet-ieee80211-radio.c, dissect_wlan_radio_phdr)   

     Rop2D      vote  
0    10.0   0.008824  
1    10.0   0.008824  
2    10.0   0.008824  
3    10.0   0.008824  
4    10.0   0.008824  
..    ...        ...  
199  10.0   0.008824  
200  10.0   0.008824  
201  10.0   0.008824  
202  10.0   0.008824  
203  13.0  -0.002262  

[204 rows x 14 columns]

Top 5:


meth  \
0  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   

       vote  rank  
0  1.487485  5.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    3373  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    4866  32   1    1    0    0.058824  0.174078  24716.0  24716.0   
2    4864  32   1    1    0    0.058824  0.174078  24716.0  24716.0   
3    4861  32   1    1    0    0.058824  0.174078  24716.0  24716.0   
4    4860  32   1    1    0    0.058824  0.174078  24716.0  24716.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
319  3514  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   
320  3512  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   
321  3505  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   
322  3502  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   
323  3405  4    0    29   1   -0.117647  0.000000  91069.0  96800.0   

                                             file                func  \
0    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
1    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
2    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
3    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
4    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
..                                            ...                 ...   
319  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
320  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
321  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
322  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
323  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   

                                                                   meth  \
0    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
1    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
2    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
3    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
4    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
..                                                                  ...   
319  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
320  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
321  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
322  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
323  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   

     Rop2D      vote  
0    10.0   0.008824  
1    11.0   0.005348  
2    11.0   0.005348  
3    11.0   0.005348  
4    11.0   0.005348  
..    ...        ...  
319  13.0  -0.002262  
320  13.0  -0.002262  
321  13.0  -0.002262  
322  13.0  -0.002262  
323  16.0  -0.007353  

[324 rows x 14 columns]


[Ochiai]
Faulty method:


meth      vote  \
0  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)  0.441942   

    rank  
0  513.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0   369   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1   355   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2   354   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3   353   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4   352   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
5   351   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
6   350   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
7   349   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
8   348   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
9   347   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
10  343   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
11  342   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
12  341   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
13  334   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
14  331   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
15  330   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
16  329   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
17  328   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
18  327   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
19  326   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
20  325   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
21  323   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
22  322   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
23  321   31   1    2    0    0.088235  0.176777  16902.0  16902.0   
24  319   31   1    2    0    0.088235  0.176777  16902.0  16902.0   

                                         file                 func  \
0   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
1   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
2   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
3   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
4   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
5   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
6   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
7   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
8   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
9   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
10  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
11  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
12  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
13  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
14  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
15  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
16  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
17  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
18  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
19  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
20  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
21  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
22  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
23  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   
24  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptCopyKey   

                                                                meth  \
0   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)   
1   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)   
2   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)   
3   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptCopyKey)   
4 

Faulty method:


meth  \
0  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   

       vote  rank  
0  1.803122  45.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    1759  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    1725  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2    1724  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3    1722  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4    1715  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
97   1477  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
98   1476  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
99   1475  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
100  1474  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
101  1467  31   1    2    0    0.088235  0.176777  16902.0  16902.0   

                                          file                         func  \
0    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
1    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
2    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
3    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
4    /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
..                                         ...                          ...   
97   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
98   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
99   /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
100  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   
101  /home/workspace/epan/crypt/dot11decrypt.c  Dot11DecryptRsna4WHandshake   

                                                                         meth  \
0    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
1    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
2    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
3    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
4    (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
..                                                                        ...   
97   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
98   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
99   (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
100  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   
101  (/home/workspace/epan/crypt/dot11decrypt.c, Dot11DecryptRsna4WHandshake)   

     RochiaiD      vote  
0    10.0      0.017678  
1    10.0      0.017678  
2    10.0      0.017678  
3    10.0      0.017678  
4    10.0      0.017678  
..    ...           ...  
97   10.0      0.017678  
98   10.0      0.017678  
99   10.0      0.017678  
100  10.0      0.017678  
101  10.0      0.017678  

[102 rows x 14 columns]

Top 5 methods:


meth  \
0  (/home/workspace/tshark.c, main)                                        
1  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)      
2  (/home/workspace/tools/lemon/lemon.c, ReportTable)                      
3  (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)             
4  (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   

       vote  rank  
0  5.022046  1.0   
1  4.622823  2.0   
2  4.287465  3.0   
3  4.150765  4.0   
4  4.116784  5.0

Top 1:


meth      vote  rank
0  (/home/workspace/tshark.c, main)  5.022046  1.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    1586  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    1438  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2    1428  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3    1320  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4    1319  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
375  845   4    0    29   1   -0.117647  0.000000  91069.0  96800.0   
376  844   4    0    29   1   -0.117647  0.000000  91069.0  96800.0   
377  837   7    0    26   1   -0.205882  0.000000  91366.0  96800.0   
378  836   7    0    26   1   -0.205882  0.000000  91366.0  96800.0   
379  835   7    0    26   1   -0.205882  0.000000  91366.0  96800.0   

                         file  func                              meth  \
0    /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
1    /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
2    /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
3    /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
4    /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
..                        ...   ...                               ...   
375  /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
376  /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
377  /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
378  /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   
379  /home/workspace/tshark.c  main  (/home/workspace/tshark.c, main)   

     RochiaiD      vote  
0    10.0      0.017678  
1    10.0      0.017678  
2    10.0      0.017678  
3    10.0      0.017678  
4    10.0      0.017678  
..    ...           ...  
375  13.0      0.000000  
376  13.0      0.000000  
377  13.0      0.000000  
378  13.0      0.000000  
379  13.0      0.000000  

[380 rows x 14 columns]

Top 2:


meth  \
0  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   

       vote  rank  
0  4.622823  2.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    3373  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    4866  32   1    1    0    0.058824  0.174078  24716.0  24716.0   
2    4864  32   1    1    0    0.058824  0.174078  24716.0  24716.0   
3    4861  32   1    1    0    0.058824  0.174078  24716.0  24716.0   
4    4860  32   1    1    0    0.058824  0.174078  24716.0  24716.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
319  3514  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   
320  3512  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   
321  3505  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   
322  3502  1    0    32   1   -0.029412  0.000000  85030.0  96800.0   
323  3405  4    0    29   1   -0.117647  0.000000  91069.0  96800.0   

                                             file                func  \
0    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
1    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
2    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
3    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
4    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
..                                            ...                 ...   
319  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
320  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
321  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
322  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   
323  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_common   

                                                                   meth  \
0    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
1    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
2    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
3    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
4    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
..                                                                  ...   
319  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
320  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
321  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
322  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   
323  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_common)   

     RochiaiD      vote  
0    10.0      0.017678  
1    11.0      0.015825  
2    11.0      0.015825  
3    11.0      0.015825  
4    11.0      0.015825  
..    ...           ...  
319  13.0      0.000000  
320  13.0      0.000000  
321  13.0      0.000000  
322  13.0      0.000000  
323  13.0      0.000000  

[324 rows x 14 columns]

Top 3:


meth      vote  rank
0  (/home/workspace/tools/lemon/lemon.c, ReportTable)  4.287465  3.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    4687  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
1    4686  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
2    4685  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
3    4683  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
4    4680  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
295  4164  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
296  4163  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
297  4162  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
298  4161  33   1    0    0    0.029412  0.171499  79913.0  79913.0   
299  4157  33   1    0    0    0.029412  0.171499  79913.0  79913.0   

                                    file         func  \
0    /home/workspace/tools/lemon/lemon.c  ReportTable   
1    /home/workspace/tools/lemon/lemon.c  ReportTable   
2    /home/workspace/tools/lemon/lemon.c  ReportTable   
3    /home/workspace/tools/lemon/lemon.c  ReportTable   
4    /home/workspace/tools/lemon/lemon.c  ReportTable   
..                                   ...          ...   
295  /home/workspace/tools/lemon/lemon.c  ReportTable   
296  /home/workspace/tools/lemon/lemon.c  ReportTable   
297  /home/workspace/tools/lemon/lemon.c  ReportTable   
298  /home/workspace/tools/lemon/lemon.c  ReportTable   
299  /home/workspace/tools/lemon/lemon.c  ReportTable   

                                                   meth  RochiaiD      vote  
0    (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
1    (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
2    (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
3    (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
4    (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
..                                                  ...   ...           ...  
295  (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
296  (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
297  (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
298  (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  
299  (/home/workspace/tools/lemon/lemon.c, ReportTable)  12.0      0.014292  

[300 rows x 14 columns]

Top 4:


meth      vote  rank
0  (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  4.150765  4.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    6676  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    6675  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2    6674  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3    6673  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4    6475  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
286  6550  4    0    29   1   -0.117647  0.000000  91069.0  96800.0   
287  6549  4    0    29   1   -0.117647  0.000000  91069.0  96800.0   
288  6497  14   0    19   1   -0.411765  0.000000  94624.0  96800.0   
289  6496  14   0    19   1   -0.411765  0.000000  94624.0  96800.0   
290  6207  14   0    19   1   -0.411765  0.000000  94624.0  96800.0   

                                             file         func  \
0    /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
1    /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
2    /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
3    /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
4    /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
..                                            ...          ...   
286  /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
287  /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
288  /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
289  /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   
290  /home/workspace/epan/dissectors/packet-tcp.c  dissect_tcp   

                                                            meth  RochiaiD  \
0    (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  10.0       
1    (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  10.0       
2    (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  10.0       
3    (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  10.0       
4    (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  10.0       
..                                                           ...   ...       
286  (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  13.0       
287  (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  13.0       
288  (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  13.0       
289  (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  13.0       
290  (/home/workspace/epan/dissectors/packet-tcp.c, dissect_tcp)  13.0       

         vote  
0    0.017678  
1    0.017678  
2    0.017678  
3    0.017678  
4    0.017678  
..        ...  
286  0.000000  
287  0.000000  
288  0.000000  
289  0.000000  
290  0.000000  

[291 rows x 14 columns]

Top 5:


meth  \
0  (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   

       vote  rank  
0  4.116784  5.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0    1942  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
1    1929  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
2    1920  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
3    1914  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
4    1913  31   1    2    0    0.088235  0.176777  16902.0  16902.0   
..    ...  ..  ..   ..   ..         ...       ...      ...      ...   
274  1516  14   0    19   1   -0.411765  0.000000  94624.0  96800.0   
275  1515  14   0    19   1   -0.411765  0.000000  94624.0  96800.0   
276  1343  10   0    23   1   -0.294118  0.000000  92304.0  96800.0   
277  1342  10   0    23   1   -0.294118  0.000000  92304.0  96800.0   
278  1341  10   0    23   1   -0.294118  0.000000  92304.0  96800.0   

                                              file                  func  \
0    /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
1    /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
2    /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
3    /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
4    /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
..                                             ...                   ...   
274  /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
275  /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
276  /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
277  /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   
278  /home/workspace/epan/dissectors/packet-http.c  dissect_http_message   

                                                                      meth  \
0    (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
1    (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
2    (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
3    (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
4    (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
..                                                                     ...   
274  (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
275  (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
276  (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
277  (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   
278  (/home/workspace/epan/dissectors/packet-http.c, dissect_http_message)   

     RochiaiD      vote  
0    10.0      0.017678  
1    10.0      0.017678  
2    10.0      0.017678  
3    10.0      0.017678  
4    10.0      0.017678  
..    ...           ...  
274  13.0      0.000000  
275  13.0      0.000000  
276  13.0      0.000000  
277  13.0      0.000000  
278  13.0      0.000000  

[279 rows x 14 columns]

# coverage hierarchy

In [25]:
pd.set_option("display.max_columns", None)
print("[fault_idx covs]")
fault_covmat_idxs = []
for fault_meth, fault_idxs_of_meth in fault_meth2idx.items():
    fault_covmat_idxs += fault_idxs_of_meth
display(covmat.astype(bool).astype(int).loc[fault_covmat_idxs])

print("[unique_covs]")

unique_covs = covmat.astype(bool).astype(int).drop_duplicates()

P = len(pass_tcs)
F = len(fail_tcs)
unique_covs["e_p"] = covmat_bool[pass_tcs].sum(axis=1)
unique_covs["e_f"] = covmat_bool[fail_tcs].sum(axis=1)
unique_covs["n_p"] = P - e_p
unique_covs["n_f"] = F - e_f

unique_covs["Sop2"] = unique_covs.apply(
    lambda row: op2(row.e_f, row.e_p, row.n_f, row.n_p), axis=1
)
unique_covs["Sochiai"] = unique_covs.apply(
    lambda row: ochiai(row.e_f, row.e_p, row.n_f, row.n_p), axis=1
)
unique_covs.sort_values("Sop2", ascending=False, inplace=True)


display(unique_covs)
pd.reset_option("display.max_columns")


[fault_idx covs]


28  \
file                                      function                    line       
/home/workspace/epan/crypt/dot11decrypt.c Dot11DecryptCopyKey         330   1    
                                                                      331   1    
                                          Dot11DecryptRsna4WHandshake 1604  1    
                                                                      1600  1    
                                                                      1590  1    
                                                                      1603  1    
                                                                      1602  1    

                                                                            6  \
file                                      function                    line      
/home/workspace/epan/crypt/dot11decrypt.c Dot11DecryptCopyKey         330   1   
                                                                      331   1   
                                          Dot11DecryptRsna4WHandshake 1604  1   
                                                                      1600  1   
                                                                      1590  1   
                                                                      1603  1   
                                                                      1602  1   

                                                                            5  \
file                                      function                    line      
/home/workspace/epan/crypt/dot11decrypt.c Dot11DecryptCopyKey         330   1   
                                                                      331   1   
                                          Dot11DecryptRsna4WHandshake 1604  1   
                                                                      1600  1   
                                                                      1590  1   
                                                                      1603  1   
                                                                      1602  1   

                                                                            15  \
file                                      function                    line       
/home/workspace/epan/crypt/dot11decrypt.c Dot11DecryptCopyKey         330   1    
                                                                      331   1    
                                          Dot11DecryptRsna4WHandshake 1604  1    
                                                                      1600  1    
                                                                      1590  1    
                                                                      1603  1    
                                                                      1602  1    

                                                                            26  \
file                                      function                    line       
/home/workspace/epan/crypt/dot11decrypt.c Dot11DecryptCopyKey         330   1    
                                                                      331   1    
                                          Dot11DecryptRsna4WHandshake 1604  1    
                                                                      1600  1    
                                                                      1590  1    
                                                                      1603  1    
                                                                      1602  1    

                                                                            16  \
file                                      function                    line       
/home/workspace/epan/crypt/dot11decrypt.c Dot11DecryptCopyKey         330   1    
                                                                      331   1    
                                          Dot11DecryptRsna4WHandshake 1

[unique_covs]


28  \
file                                                   function                           line       
/home/workspace/epan/prefs.c                           register_preference                1039  0    
/home/workspace/epan/proto.c                           find_protocol_by_id                7298  0    
/home/workspace/epan/addr_resolv.c                     parse_ether_address                1388  0    
/home/workspace/epan/prefs.c                           register_preference                1003  0    
/home/workspace/epan/oids.c                            check_num_oid                      893   0    
/home/workspace/epan/crypt/dot11decrypt.c              Dot11DecryptRsnaPwd2PskStep        2427  0    
/home/workspace/epan/wmem/wmem_map.c                   wmem_strong_hash                   428   0    
/home/workspace/epan/proto.c                           proto_register_field_common        7575  1    
/home/workspace/epan/wmem/wmem_map.c                   wmem_strong_hash                   429   1    
/home/workspace/epan/dissectors/packet-knxip_decrypt.c encrypt_block                      35    1    
/home/workspace/epan/wmem/wmem_core.c                  wmem_alloc                         40    1    
/home/workspace/epan/wmem/wmem_allocator_block.c       wmem_block_split_free_chunk        563   1    
/home/workspace/epan/dissectors/packet-xmcp.c          dissect_xmcp_heur                  1045  1    
/home/workspace/epan/dissectors/packet-gluster_pmap.c  proto_register_gluster_pmap        174   1    
/home/workspace/epan/wmem/wmem_tree.c                  wmem_tree_remove_string            628   0    
/home/workspace/epan/proto.c                           find_protocol_by_id                7305  0    
/home/workspace/epan/wmem/wmem_tree.c                  wmem_tree_destroy_cb               226   0    
/home/workspace/wsutil/strtoi.c                        ws_basestrtou64                    154   0    
/home/workspace/epan/wmem/wmem_tree.c                  wmem_tree_lookup_string            621   0    
/home/workspace/epan/dissectors/packet-rpc.c           dissect_rpc_fragment               3203  0    
/home/workspace/wsutil/json_dumper.c                   json_puts_string                   71    0    
/home/workspace/epan/dissectors/packet-dns.c           dissect_dns_common                 3839  1    
/home/workspace/wiretap/wtap.c                         wtap_free_idb_info                 154   1    
/home/workspace/rawshark.c                             log_func_ignore                    213   1    
/home/workspace/epan/dissectors/packet-tcp.c           check_follow_fragments             984   1    
/home/workspace/wiretap/pcapng.c                       pcapng_write_enhanced_packet_block 3253  1    
/home/workspace/extcap/androiddump.c                   main                               2593  1    
/home/workspace/epan/dissectors/packet-asterix.c       dissect_asterix                    9909  1    
/home/workspace/epan/wmem/wmem_tree.c                  wmem_tree_lookup32_le              572   1    
/home/workspace/epan/tvbuff.c                          tvb_find_tvb                       4018  1    
/home/workspace/epan/ftypes/ftype-time.c               cmp_ne                             40    1    
                                                       cmp_eq                             34    1    
/home/workspace/epan/dissectors/packet-tr.c            dissect_tr                         361   1    
/home/workspace/epan/ftypes/ftype-bytes.c              bytes_from_string                  221   1    
/home/workspace/epan/dfilter/scanner.l                 NaN                                263   1    
                                                                                          128   1    
/home/workspace/epan/dissectors/packet-nfs.c           dissect_fhandle_data_unknown       2300  1    
/home/workspace/epan/ftypes/ftype-integer.c            ipxnet_from_unparsed               506   1    
/home/workspace/

# 감상

Coverage vector 상에선 unique한 점이 있으나, max aggregation에서 faulty method의 tie 수가 1051개 (rank는 그 보다 잘하는 17개 밑이라 1068); test가 부족하다.
